### PR TITLE
Use overlay

### PR DESCRIPTION
**Pull request by anitagraser**
I suggest to use the much more straightforward overlay approach instead of manually selecting a single df row's geometry.

### CODE REVIEW

File: 04-spatial-operations.qmd [hunk 1]
```diff
@@ -21,11 +21,13 @@ plt.rcParams["figure.figsize"] = (5, 5)
 Packages...
 
 ```{python}
-import numpy as np
 import geopandas as gpd
+import numpy as np
```
Comment (Robinlovelace) [inline 04-spatial-operations.qmd:25]: Any reason for the ordering?

File: 04-spatial-operations.qmd [hunk 2]
```diff
@@ -71,8 +71,7 @@ Spatial subsetting...
 
 ```{python}
 canterbury = nz[nz["Name"] == "Canterbury"]
-sel = nz_height.intersects(canterbury["geometry"].iloc[0])
-canterbury_height = nz_height[sel]
+canterbury_height = nz_height.overlay(canterbury, how='intersection')
```
Comment (Robinlovelace) [inline 04-spatial-operations.qmd:74]: This looks neater to me for sure.

File: 04-spatial-operations.qmd [hunk 3]
```diff
@@ -85,8 +84,7 @@ canterbury_height.plot(ax=base, color="None", edgecolor="red");
 Spatial subsetting 2...
 
 ```{python}
-sel = nz_height.disjoint(canterbury["geometry"].iloc[0])
-non_canterbury_height = nz_height[sel]
+non_canterbury_height = nz_height.overlay(canterbury, how='difference')
```
Comment (Robinlovelace) [inline 04-spatial-operations.qmd:87]: :+1:

**Comment by anitagraser** [comment]
This should have been two separate PRs, sorry. Well, let me know what you think.

File: 04-spatial-operations.qmd (hunk 3 above)
Comment (Robinlovelace) [reply 04-spatial-operations.qmd:87]: Maybe we should add more words explaining this. But that can happen post merge.

**Review by Robinlovelace — APPROVED** [review]

**Comment by Robinlovelace** [comment]
Looks great to me Anita, there are many ways to do things and it's tempting to show various ways. But given this book's focus on getting things done quickly and easily I think showing people the easiest way is best and your approach is definitely better. One idea: worth adding a footnote or something showing alternative approaches? Sometimes it's useful to have the indexes of the entities as in the previous `sel` object. Very happy for this to be merged in any case.

File: 04-spatial-operations.qmd (hunk 1 above)
Comment (anitagraser) [reply 04-spatial-operations.qmd:25]: Just alphabetical.

**Comment by anitagraser** [comment]
>  One idea: worth adding a footnote or something showing alternative approaches? 

Something along the following line?

In case you need the indexes of the intersecting features, you can use: 
`nz_height.intersects(canterbury["geometry"].iloc[0])`

Actually, this seems to be covered in the following section on topological relationships: 

image

**Comment by Robinlovelace** [comment]
> Something along the following line?
> 
> In case you need the indexes of the intersecting features, you can use: `nz_height.intersects(canterbury["geometry"].iloc[0])`
> 
> Actually, this seems to be covered in the following section on topological relationships:

:+1: already covered. Less can be more!

**Comment by michaeldorman** [comment]
Thanks! I agree that `nz_height.overlay(canterbury, how='difference')` is more straightforward in the particular case of spatial subsetting of points, but if one needs to subset lines/polygons then `.overlay` refers to a different operation: "clipping" rather than selecting. If we switch the point example to `.overlay` then IMHO we also need to include a line/polygon example, so that the reader is not confused if trying to implement `.overlay` to select lines/polygons and seeing an unexprected result. Will be happy to hear what you think, thanks.

**Comment by michaeldorman** [comment]
Alternatively, maybe we can keep both code sections and specify that the second (shorter one) is applicable to points, while the first is applicable to all geometry types, and then we can avoid adding a new line/polygon example.

**Comment by Robinlovelace** [comment]
> if one needs to subset lines/polygons then `.overlay` refers to a different operation: "clipping" rather than selecting. If we switch the point example to `.overlay` then IMHO we also need to include a line/polygon example, so that the reader is not confused

I didn't realise this. It's an important distinction that should for sure be documented. So `.overlay()` filters (removes features with geometries totally outside the subsetting object) and changes geometries (of features partly intersecting with the subsetting object)? Both are important but 'pure subsetting' without changing geometries should go first logically I guess.

**Comment by michaeldorman** [comment]
`x.overlay(y, how='intersection')` is same as `sf::st_intersection(x,y)` in R. 

Therefore, for points it's equivalent to spatial subsetting, but for other geometry types it isn't.

**Comment by Robinlovelace** [comment]
And is therefore most relevant in the geometry chapter as it changes the geometry of objects.

**Comment by anitagraser** [comment]
In QGIS terminology, spatial subsetting would be "select by location". Related concepts are select by attribute and select by expression. Could be useful to mention, since readers may recognize these terms.

**Comment by Robinlovelace** [comment]
Big thumbs up to that, QGIS is probably the biggest open source geo project in terms of n. users. With many thousands of people discovering GIS concepts through QGIS it's well worth mentioning QGIS equivalents. To add another term, in 'tidyverse' and [SQL syntax](https://swcarpentry.github.io/sql-novice-survey/03-filter/) I think it would be called a spatial filter. Another important point of reference is PostGIS, which has great docs that could be good to cross-reference. Cannot see what they call a spatial filter/subset/selection here: https://postgis.net/docs/using_postgis_query.html tho...

**Pull request merged**